spec.html

```diff
@@ -282,142 +373,130 @@ <h1>%Iterator.prototype%.constructor</h1>
-            1. IfAbruptCloseIterator(_iterated_, _selected_).
-            1. If ToBoolean(_selected_) is *true*,
-              1. Set _lastValue_ to Yield(_value_).
-              1. IfAbruptCloseIterator(_iterated_, _lastValue_).
+          1. Let _closure_ be a new Abstract Closure with no parameters that captures _iterated_ and _filterer_ and performs the following steps when called:
+            1. Let _lastValue_ be *undefined*.
+            1. Repeat,
+              1. Let _next_ be ? IteratorStep(_iterated_, _lastValue_).
+              1. If _next_ is *false*, return *undefined*.
+              1. Let _value_ be ? IteratorValue(_next_).
+              1. Let _selected_ be Call(_filterer_, *undefined*, &laquo; _value_ &raquo;).
+              1. IfAbruptCloseIterator(_iterated_, _selected_).
+              1. If ToBoolean(_selected_) is *true*,
+                1. Set _lastValue_ to Yield(_value_).
+                1. IfAbruptCloseIterator(_iterated_, _lastValue_).
+          1. Return ? CreateBuiltinGeneratorInstance(_closure_, ~Iterator Helper~, %IteratorHelperPrototype%).
         </emu-alg>
       </emu-clause>
 
       <emu-clause id="sec-iteratorprototype.take">
         <h1>%Iterator.prototype%.take ( _limit_ )</h1>
-        <p>%Iterator.prototype%.take is a built-in generator function which, when called, performs the following prelude steps:</p>
         <emu-alg>
           1. Let _iterated_ be ? GetIteratorDirect(*this* value).
           1. Let _remaining_ be ? ToInteger(_limit_).
           1. If _remaining_ &lt; 0, throw a *RangeError* exception.
-        </emu-alg>
-        <p>The body of %Iterator.prototype%.take is composed of the following steps</p>
-        <emu-alg>
-          1. Let _lastValue_ be *undefined*.
-          1. Repeat,
-            1. If _remaining_ is 0, then
-              1. Return ? IteratorClose(_iterated_, NormalCompletion(*undefined*)).
-            1. Set _remaining_ to _remaining_ - 1.
-            1. Let _next_ be ? IteratorStep(_iterated_, _lastValue_).
-            1. If _next_ is *false*, return *undefined*.
-            1. Set _lastValue_ to Yield(? IteratorValue(_next_)).
-            1. IfAbruptCloseIterator(_iterated_, _lastValue_).
+          1. Let _closure_ be a new Abstract Closure with no parameters that captures _iterated_ and _remaining_ and performs the following steps when called:
+            1. Let _lastValue_ be *undefined*.
+            1. Repeat,
+              1. If _remaining_ is 0, then
+                1. Return ? IteratorClose(_iterated_, NormalCompletion(*undefined*)).
+              1. Set _remaining_ to _remaining_ - 1.
+              1. Let _next_ be ? IteratorStep(_iterated_, _lastValue_).
+              1. If _next_ is *false*, return *undefined*.
+              1. Set _lastValue_ to Yield(? IteratorValue(_next_)).
+              1. IfAbruptCloseIterator(_iterated_, _lastValue_).
```

Comment:
if `_next_` or `IteratorValue(_next_)` are abrupt completions, the iterator seems like it won't be closed? same elsewhere.

Reply:
My understanding is that if the abrupt completion comes from the source iterator directly (i.e. it's next method threw or the iterator result it returned was invalid), it's considered to be an invalid iterator so it doesn't get closed. All of the methods follow that same pattern, IDK if that behaviour's desirable or not.

Reply:
Right. If I understand correctly, this is how it works with `for` loops and other kinds of iteration in the spec, so we're just following existing conventions here.